File: day03/solution.py
from utils.file import read_file_content
from utils.list import deep_copy


def solve_part1(input: str) -> int:
    lines = input.split("\n")
    total = len(lines)

    r1 = ""
    r2 = ""
    for i in range(len(lines[0])):
        ones = len([l for l in lines if l[i] == '1'])
        if ones >= total/2:
            r1 += '1'
            r2 += '0'
        else:
            r1 += '0'
            r2 += '1'

    return int(r1, 2)*int(r2, 2)


def solve_part2(input: str) -> int:
    lines = input.split("\n")

    oxList = deep_copy(lines)
    coList = deep_copy(lines)

    for i in range(len(lines[0])):
        o1 = len([l for l in oxList if l[i] == '1'])
        o2 = len([l for l in coList if l[i] == '1'])

        if len(oxList) > 1:
            if o1 >= len(oxList)/2:
                oxList = [l for l in oxList if l[i] == '1']
            else:
                oxList = [l for l in oxList if l[i] == '0']

        if len(coList) > 1:
            if o2 >= len(coList)/2:
                coList = [l for l in coList if l[i] == '0']
            else:
                coList = [l for l in coList if l[i] == '1']

    return int(oxList[0], 2) * int(coList[0], 2)


def test_part1():
    input = read_file_content("inputs/test")
    answer = int(read_file_content("inputs/ans1"))

    result = solve_part1(input)
    if result == answer:
        print("Test successful")
    else:
        print("Test unsuccessful: " + str(result) + ", expected: " + str(answer))


def test_part2():
    input = read_file_content("inputs/test")
    answer = int(read_file_content("inputs/ans2"))

    result = solve_part2(input)
    if result == answer:
        print("Test successful")
    else:
        print("Test unsuccessful: " + str(result) + ", expected: " + str(answer))


if __name__ == '__main__':

    input = read_file_content("inputs/input")

    print(" --- Part 1 --- ")
    test_part1()
    print("Part 1 result:\t" + str(solve_part1(input)))

    print("\n --- Part 2 ---")
    test_part2()
    print("Part 2 result:\t" + str(solve_part2(input)))

# I knew there was a better way to count/filter lists in Python, but in the moment did not think of it.
# So once it was done i wanted to refresh my memory, and implemented it with better list manipulation.
# These are the original methods as backup

# def solve_part1(input: str) -> int:
#     lines = input.split("\n")
#     numLines = len(lines)
#
#     occ = [0]*len(lines[0])
#     for l in lines:
#         for i in range(len(l)):
#             if l[i] == '1':
#                 occ[i] += 1
#
#     r1 = ""
#     r2 = ""
#     for o in occ:
#         if o >= numLines/2:
#             r1 += '1'
#             r2 += '0'
#         else:
#             r1 += '0'
#             r2 += '1'
#
#     num1 = int(r1, 2)
#     num2 = int(r2, 2)
#
#     return num1*num2

# def solve_part2(input: str) -> int:
#     lines = input.split("\n")
#
#     oxList = deep_copy(lines)
#     coList = deep_copy(lines)
#
#     oxResult = None
#     coResult = None
#
#     for i in range(len(lines[0])):
#         o1 = count_ones(oxList, i)
#         o2 = count_ones(coList, i)
#         oxToRemove = []
#         coToRemove = []
#         if o1 >= len(oxList)/2:
#             for l in oxList:
#                 if l[i] != '1':
#                     oxToRemove.append(l)
#         else:
#             for l in oxList:
#                 if l[i] != '0':
#                     oxToRemove.append(l)
#
#         if o2 >= len(coList)/2:
#             for l in coList:
#                 if l[i] != '0':
#                     coToRemove.append(l)
#         else:
#             for l in coList:
#                 if l[i] != '1':
#                     coToRemove.append(l)
#
#         if oxResult == None:
#             for l in oxToRemove:
#                 oxList.remove(l)
#             if len(oxList) == 1:
#                 oxResult = oxList[0]
#
#         if coResult == None:
#             for l in coToRemove:
#                 coList.remove(l)
#             if len(coList) == 1:
#                 coResult = coList[0]
#
#     coVal = int(coResult, 2)
#     oxVal = int(oxResult, 2)
#
#     return coVal*oxVal

File: day03/solution.go
package main

import (
	"aoc21/utils"
	"fmt"
	"strconv"
	"strings"
)

const DAY = "03"


func solvePart1 (input string) int {
	var lines = strings.Split(input, "\r\n")
	var length = len(lines)

	var occurrences []int
	for i := 0; i < len(lines[0]); i++ {
		occurrences = append(occurrences, 0)
	}

	for _, line := range lines {
		for i := 0; i < len(line); i++{
			if line[i] == '1' {
				occurrences[i] += 1
			}
		}
	}

	var gamma = ""
	var epsilon = ""
	for _, occ := range occurrences {
		if occ > length/2 {
			gamma += "1"
			epsilon += "0"
		} else {
			gamma += "0"
			epsilon += "1"
		}
	}

	var gammaVal,_ = strconv.ParseInt(gamma, 2, 16)
	var epsilonVal,_ = strconv.ParseInt(epsilon, 2, 16)

	return int(gammaVal * epsilonVal)
}

func solvePart2 (input string) int {
	var lines = strings.Split(input, "\r\n")

	var oxyList = map[string]bool{}
	var co2List = map[string]bool{}
	for _,line := range lines {
		oxyList[line] = true
		co2List[line] = true
	}

	for i := 0; i < len(lines[0]); i++ {
		var oxyListLength = len(oxyList)
		var co2ListLength = len(co2List)

		// Count 1's on both lists
		var oxyOccurrences = 0
		for key := range oxyList {
			if key[i] == '1' {
				oxyOccurrences += 1
			}
		}
		var co2Occurrences = 0
		for key := range co2List {
			if key[i] == '1' {
				co2Occurrences += 1
			}
		}

		// Delete "wrong" bits from the oxygen list
		if oxyListLength > 1 {
			var oxyDeleteList = map[string]bool{}
			if float64(oxyOccurrences) >= float64(oxyListLength)/2 {
				for key:= range oxyList {
					if key[i] == '0' {
						oxyDeleteList[key] = true
					}
				}
			} else {
				for key := range oxyList {
					if key[i] == '1' {
						oxyDeleteList[key] = true
					}
				}
			}
			for key := range oxyDeleteList {
				delete(oxyList, key)
			}
		}

		// Delete "wrong" bits from the CO2 list
		if co2ListLength > 1 {
			var co2DeleteList = map[string]bool{}
			if float64(co2Occurrences) >= float64(co2ListLength)/2 {
				for key := range co2List {
					if key[i] == '1' {
						co2DeleteList[key] = true
					}
				}
			} else {
				for key := range co2List {
					if key[i] == '0' {
						co2DeleteList[key] = true
					}
				}
			}
			for key := range co2DeleteList {
				delete(co2List, key)
			}
		}
	}

	var oxyValue, co2Value int64
	for key := range oxyList {
		oxyValue,_ = strconv.ParseInt(key, 2, 32)
	}
	for key:= range co2List {
		co2Value,_ = strconv.ParseInt(key, 2, 32)
	}

	return int(oxyValue * co2Value)
}


func testPart1 () bool {
	var input = utils.ReadFileContent("day" + DAY + "/inputs/test")
	var answer,_ = strconv.Atoi(utils.ReadFileContent("day" + DAY + "/inputs/ans1"))

	var result = solvePart1(input)
	if result != answer {
		fmt.Println("Test unsuccessful: " + strconv.Itoa(result) + ", expected: " + strconv.Itoa(answer))
	}
	return result == answer
}

func testPart2 () bool {
	var input = utils.ReadFileContent("day" + DAY + "/inputs/test")
	var answer,_ = strconv.Atoi(utils.ReadFileContent("day" + DAY + "/inputs/ans2"))

	var result = solvePart2(input)
	if result != answer {
		fmt.Println("Test unsuccessful: " + strconv.Itoa(result) + ", expected: " + strconv.Itoa(answer))
	}
	return result == answer
}

func main () {
	var input = utils.ReadFileContent("day" + DAY + "/inputs/input")

	fmt.Println(" ~~~ Advent of Code 2021, Day " + DAY + " ~~~")

	fmt.Println(" --- Part 1 --- ")
	var test = testPart1()
	if test {
		fmt.Println("Test Successful")
		fmt.Println("Part 1 result:\t" + strconv.Itoa(solvePart1(input)))
	}

	fmt.Println("\n --- Part 2 --- ")
	test = testPart2()
	if test {
		fmt.Println("Test Successful")
		fmt.Println("Part 2 result:\t" + strconv.Itoa(solvePart2(input)))
	}
}
